src/Oscoin/P2P/Types.hs

```diff
@@ -50,19 +59,90 @@ import qualified Codec.Serialise as CBOR
 import qualified Codec.Serialise.Decoding as CBOR
 import qualified Codec.Serialise.Encoding as CBOR
 import           Control.Monad.Fail (fail)
+import           Control.Monad.Trans.Writer (execWriterT, tell)
 import           Data.Aeson (FromJSON, ToJSON)
 import           Data.Char (isAlphaNum)
 import           Data.Hashable (Hashable(..))
 import           Data.IP (IP(..))
 import           Data.Profunctor (Profunctor(dimap))
+import           Data.String (IsString(..))
 import qualified Data.Text as T
 import           Data.Vector (Vector)
 import qualified Data.Vector as V
+import qualified Data.Vector.Unboxed as V.Unboxed
 import           Formatting as F
 import qualified Generics.SOP as SOP
 import qualified Network.DNS as DNS
 import           Network.Socket (PortNumber, SockAddr)
 import qualified Network.Socket as Network
+import           System.Random (RandomGen, randomR)
+
+-- | The name of the overlay network.
+--
+-- This is, in a way, the \"physical\" network a node joins, although physical
+-- isolation is not implied. The overlay network typically matches the
+-- \"logical\" oscoin network as defined by "Oscoin.Configuration.Network".
+-- Additionally, ad-hoc networks ('Somenet'') can be created, which map to the
+-- logical "Oscoin.Configuration.Devnet".
+--
+-- Ad-hoc networks must be valid domain name labels, and can only be created
+-- using 'readNetwork', 'readNetworkText', or 'randomNetwork'. To allow
+-- exhaustive pattern matching, the view pattern 'Somenet' is exported.
+--
+data Network =
+      Mainnet
+    | Testnet
+    | Devnet
+    | Somenet' Text
+    deriving (Eq, Show)
+
+pattern Somenet :: Text -> Network
+pattern Somenet x <- Somenet' x
+
+{-# COMPLETE Mainnet, Testnet, Devnet, Somenet #-}
+
+instance IsString Network where
+    fromString = either (panic . toS) identity . readNetwork
+
+renderNetwork :: Network -> Text
+renderNetwork Mainnet      = "mainnet"
+renderNetwork Testnet      = "testnet"
+renderNetwork Devnet       = "devnet"
+renderNetwork (Somenet' x) = x
+
+readNetwork :: String -> Either String Network
+readNetwork = readNetworkText . toS
+
+readNetworkText :: Text -> Either String Network
+readNetworkText "mainnet" = pure Mainnet
+readNetworkText "testnet" = pure Testnet
+readNetworkText "devnet"  = pure Devnet
+readNetworkText t
+  | T.length t > 63         = Left "Network name longer than 63 characters"
+  | Just c <- invalidChar t = Left $ "Invalid character in network name: " <> [c]
+  | otherwise               = Right $ Somenet' t
+  where
+    invalidChar = T.find (`elem` ("./:" :: String))
```

Comment:
It is unclear what a “valid domain name label” is. This code seem to imply that it is anything that does not contain the characters `./:`. The tests however generate only alphanumeric network names.

I’d suggest to document what a valid network name is explicitly and make the test cover the full range.. I’d prefer to use just alphanumeric names (including dashes).